src/main/java/hudson/plugins/git/GitSCM.java

```diff
@@ -1148,6 +1156,30 @@ private void retrieveChanges(Run build, GitClient git, TaskListener listener) th
         }
     }
 
+    private boolean determineRedundantFetch(CloneOption option, RemoteConfig rc) {
+        List<RefSpec> initialFetchRefSpecs = rc.getFetchRefSpecs();
+        boolean isDefaultRefspec = true; // default refspec is any refspec with "refs/heads/" mapping
+        boolean removeSecondFetch = true;
+        if (initialFetchRefSpecs != null) {
+            for (RefSpec ref : initialFetchRefSpecs) {
+                if (!ref.toString().contains("refs/heads")) {
+                    isDefaultRefspec = false; // if refspec is not of default type, preserve second fetch
+                }
+            }
+            if (option == null) {
+                removeSecondFetch = isDefaultRefspec;
+            } else {
+                if (!option.isHonorRefspec()) {
+                    removeSecondFetch = isDefaultRefspec;
+                } else {
+                    removeSecondFetch = true; // avoid second fetch call if honor refspec is enabled
+                }
+            }
+        }
+        // if initial fetch refspec contains "refs/heads/*" (default refspec), ignore the second fetch call
+        return removeSecondFetch;
```

Comment:
if `initialFetchRefSpecs` is null, then you're removing the second fetch. At this moment, I cannot think about a situation when it will be null, but ...

Reply:
I added a test that passed in a null refspec but the UserRemoteConfig constructor converts nulls to empty strings.  I think that is unreachable code due to the UserRemoteConfig use of `fixEmpty()`.

Reply:
According to my assumptions, the need to add the null check was to check for empty refspecs, as Mark as correctly pointed out, the `fixEmpty()` converts any empty refspec into a null refspec.

If the UserRemoteConfig was not fixing empty refspecs, the logic presented by me would gladly pass `empty` refspecs as a narrow refspec and will not avoid the second fetch. But since it is there, it should not be a concern for us.